common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -803,21 +828,25 @@ public static Inet6Address getByName(CharSequence ip, boolean ipv4Mapped) {
                 break;
             case '.':
                 ++ipv4Seperators;
-                if (i - begin > IPV4_MAX_CHAR_BETWEEN_SEPARATOR
+                tmp = i - begin; // tmp is the length of the current segment.
+                if (tmp > IPV4_MAX_CHAR_BETWEEN_SEPARATOR
+                        || begin < 0
                         || ipv4Seperators > IPV4_SEPARATORS
                         || (ipv6Seperators > 0 && (currentIndex + compressLength < 12))
                         || i + 1 >= ipLength
                         || currentIndex >= bytes.length
-                        || begin < 0
-                        || (begin == 0 && (i == 3 && (!isValidNumericChar(ip.charAt(2)) ||
-                                                      !isValidNumericChar(ip.charAt(1)) ||
-                                                      !isValidNumericChar(ip.charAt(0))) ||
-                                           i == 2 && (!isValidNumericChar(ip.charAt(1)) ||
-                                                      !isValidNumericChar(ip.charAt(0))) ||
-                                           i == 1 && !isValidNumericChar(ip.charAt(0))))) {
+                        || ipv4Seperators == 1 && (!ipv4Mapped ||
+                                             // We also parse pure IPv4 addresses as IPv4-Mapped for ease of use.
+                                                   currentIndex != 0 && !isValidIPv4Mapped(bytes, currentIndex))
+                        || (tmp == 3 && (!isValidNumericChar(ip.charAt(i - 1)) ||
+                                         !isValidNumericChar(ip.charAt(i - 2)) ||
+                                         !isValidNumericChar(ip.charAt(i - 3))) ||
+                            tmp == 2 && (!isValidNumericChar(ip.charAt(i - 1)) ||
+                                         !isValidNumericChar(ip.charAt(i - 2))) ||
+                            tmp == 1 && !isValidNumericChar(ip.charAt(i - 1)))) {
```

Comment:
Good catch. The validation after we see the first `ipv4Separators` is done in the default case below.